File: php/lib/steam/community/GameLeaderboard.php
<?php
/**
 * This code is free software; you can redistribute it and/or modify it under
 * the terms of the new BSD License.
 *
 * Copyright (c) 2011, Nicholas Hastings
 *
 * @license http://www.opensource.org/licenses/bsd-license.php New BSD License
 */

require_once STEAM_CONDENSER_PATH . 'steam/community/GameLeaderboardEntry.php';
require_once STEAM_CONDENSER_PATH . 'steam/community/SteamId.php';

/**
 * The GameLeaderboard class represents a game leaderboard for a specific game
 *
 * @author     Nicholas Hastings
 * @package    steam-condenser
 * @subpackage community
 */
class GameLeaderboard
{
    // From OpenSteamworks
    const LEADERBOARD_DISPLAY_TYPE_NONE         = 0;
    const LEADERBOARD_DISPLAY_TYPE_NUMERIC      = 1;
    const LEADERBOARD_DISPLAY_TYPE_SECONDS      = 2;
    const LEADERBOARD_DISPLAY_TYPE_MILLISECONDS = 3;

    const LEADERBOARD_SORT_METHOD_NONE = 0;
    const LEADERBOARD_SORT_METHOD_ASC  = 1; // top-score is lowest number
    const LEADERBOARD_SORT_METHOD_DESC = 2; // top-score is highest number
    //

    /**
     * @var array
     */
    private static $leaderboards = array();

    /**
     * @var int
     */
    protected $id;

    /**
     * @var string
     */
    protected $url;

    /**
     * @var string
     */
    protected $name;

    /**
     * @var int
     */
    protected $entryCount;

    /**
     * @var int
     */
    protected $sortMethod;

    /**
     * @var int
     */
    protected $displayType;

    /**
     * Returns the GameLeaderboard matching the given Leaderboard Id.
     *
     * @return GameLeaderboard or FALSE if no match
     */
    public static function getLeaderboard($gameName, $id) {
        $getter = is_string($id) ? 'getName' : 'getId';

        foreach(self::getLeaderboards($gameName) as $board) {
            if($id == $board->$getter()) {
                return $board;
            }
        }
    }

    /**
     * Returns an array containing all of the game's leaderboards
     */
    public static function getLeaderboards($gameName) {
        if(!array_key_exists($gameName, self::$leaderboards)) {
            self::loadGameLeaderboards($gameName);
        }

        return self::$leaderboards[$gameName];
    }

    private static function loadGameLeaderboards($gameName) {
        $url = "http://steamcommunity.com/stats/$gameName/leaderboards/?xml=1";
        $boardsData = new SimpleXMLElement(file_get_contents($url));

        if(!empty($boardsData->error)) {
            throw new SteamCondenserException((string) $boardsData->error);
        }

        self::$leaderboards[$gameName] = array();
        foreach($boardsData->leaderboard as $boardData) {
            self::$leaderboards[$gameName][] = new GameLeaderboard($boardData);
        }
    }

    /**
     * Creates a GameLeaderboard object
     * @param SimpleXMLElement $boardData
     */
    public function __construct(SimpleXMLElement $boardData) {
        $this->url         = (string) $boardData->url;
        $this->id          = (int)    $boardData->lbid;
        $this->name        = (string) $boardData->name;
        $this->entryCount  = (int)    $boardData->entries;
        $this->sortMethod  = (int)    $boardData->sortmethod;
        $this->displayType = (int)    $boardData->displaytype;
    }

    /**
     * @return String
     */
    public function getName() {
        return $this->name;
    }

    /**
     * @return int
     */
    public function getId() {
        return $this->id;
    }

    /**
     * @return int
     */
    public function getEntryCount() {
        return $this->entryCount;
    }

    /**
     * @return int
     */
    public function getSortMethod() {
        return $this->sortMethod;
    }

    /**
     * @return int
     */
    public function getDisplayType() {
        return $this->displayType;
    }

    /**
     * Returns the GameLeaderboardEntry on this GameLeaderboard for given steamId
     * @param steam64 string OR fetched SteamId object
     * @return GameLeaderboardEntry or FALSE if player is not on leaderboard
     */
    public function getEntryForSteamId($steamId) {
        // We'll let users pass SteamId class or steamid64 string
        if(is_object($steamId)) {
            $id = $steamId->getSteamId64();
        } else {
            $id = $steamId;
        }

        $fullurl = sprintf('%s&steamid=%s', $this->url, $id);
        $xml = new SimpleXMLElement(file_get_contents($fullurl));

        if(!empty($xml->error)) {
            throw new SteamCondenserException((string) $xml->error);
        }

        foreach($xml->entries->entry as $entryData) {
            if($entryData->steamid != $id) {
                continue;
            }

            return new GameLeaderboardEntry($entryData, $this);
        }

        return false;
    }

    /**
     * Returns an array of GameLeaderboardEntrys on this GameLeaderboard for given steamId
     * and friends of given steamId
     * @param steam64 string OR fetched SteamId object
     * @return array
     */
    public function getEntryForSteamIdFriends($steamId) {
        // We'll let users pass SteamId class or steamid64 string
        if(is_object($steamId)) {
            $id = $steamId->getSteamId64();
        } else if (is_numeric($steamId)) {
            $id = $steamId;
        } else {
            throw new SteamCondenserException('Invalid steam id. You must pass a steamId64 or a fetched SteamId object.');
        }

        $fullurl = sprintf('%s&steamid=%s', $this->url, $id);
        $xml = new SimpleXMLElement(file_get_contents($fullurl));

        if(!empty($xml->error)) {
            throw new SteamCondenserException((string) $xml->error);
        }

        $entries = array();
        foreach($xml->entries->entry as $entryData) {
            $rank = (int) $entryData->rank;
            $entries[$rank] = new GameLeaderboardEntry($entryData, $this);
        }

        return $entries;
    }

    /**
     * Returns an array of GameLeaderboardEntrys on this GameLeaderboard for given rank range.
     * Range is inclusive. Currently, a maximum of 5001 entries can be returned in a single request.
     * @param int $first
     * @param int $last
     * @return array
     */
    public function getEntryRange($first, $last) {
        if($last < $first) {
            throw new SteamCondenserException('First entry must be prior to last entry for leaderboard entry lookup.');
        }
        if(($last - $first) > 5000) {
            throw new SteamCondenserException('Leaderboard entry lookup is currently limited to a maximum of 5001 entries per request.');
        }

        $fullurl = sprintf('%s&start=%d&end=%d', $this->url, $first, $last);
        $xml = new SimpleXMLElement(file_get_contents($fullurl));

        if(!empty($xml->error)) {
            throw new SteamCondenserException((string) $xml->error);
        }

        $entries = array();
        foreach($xml->entries->entry as $entryData) {
            $rank = (int) $entryData->rank;
            $entries[$rank] = new GameLeaderboardEntry($entryData, $this);
        }

        return $entries;
    }
}
?>

File: ruby/lib/steam/community/game_leaderboard.rb
# This code is free software; you can redistribute it and/or modify it under
# the terms of the new BSD License.
#
# Copyright (c) 2011, Sebastian Staudt

require 'steam/community/game_leaderboard_entry';
require 'steam/community/steam_id';

# The GameLeaderboard class represents a single leaderboard for a specific game
#
# @author Sebastian Staudt
class GameLeaderboard

  LEADERBOARD_DISPLAY_TYPE_NONE         = 0
  LEADERBOARD_DISPLAY_TYPE_NUMERIC      = 1
  LEADERBOARD_DISPLAY_TYPE_SECONDS      = 2
  LEADERBOARD_DISPLAY_TYPE_MILLISECONDS = 3

  LEADERBOARD_SORT_METHOD_NONE = 0
  LEADERBOARD_SORT_METHOD_ASC  = 1
  LEADERBOARD_SORT_METHOD_DESC = 2

  @@leaderboards = {}

  # Returns the display type of the scores on this leaderboard
  #
  # @return [Fixnum] The display type of the scores
  attr_reader :display_type

  # Returns the number of entries on this leaderboard
  #
  # @return [Fixnum] The number of entries on this leaderboard
  attr_reader :entry_count

  # Returns the ID of the leaderboard
  #
  # @return [Fixnum] The ID of the leaderboard
  attr_reader :id

  # Returns the name of the leaderboard
  #
  # @return [String] The name of the leaderboard
  attr_reader :name

  # Returns the method that is used to sort the entries on the leaderboard
  #
  # @return [Fixnum] The sort method
  attr_reader :sort_method

  # Returns the leaderboard for the given game and leaderboard ID or name
  #
  # @param [String] game_name The short name of the game
  # @param [Fixnum, String] id The ID or name of the leaderboard to return
  # @return [GameLeaderboard] The matching leaderboard if available
  def self.leaderboard(game_name, id)
    leaderboards = self.leaderboards game_name

    if id.is_a? Fixnum
      leaderboards[id]
    else
      leaderboards.each_value do |board|
        return board if board.name == id
      end
    end
  end

  # Returns an array containing all of a game's leaderboards
  #
  # @param [String] game_name The name of the game
  # @return [Array<GameLeaderboard>] The leaderboards for this game
  def self.leaderboards(game_name)
      unless @@leaderboards.key? game_name
        self.load_game_leaderboards game_name
      end

    @@leaderboards[game_name]
  end

  # Returns the entry on this leaderboard for the user with the given SteamID
  #
  # @param [Fixnum, SteamId] steam_id The 64bit SteamID or the `SteamId` object
  #        of the user
  # @return [GameLeaderboardEntry] The entry of the user if available
  # raise [SteamCondenserException] if an error occurs while fetching the
  #         leaderboard
  def entry_for_steam_id(steam_id)
    steam_id = steam_id.steam_id64 if steam_id.is_a? SteamId

    url = "#{@url}&steamid=#{steam_id}"
    xml = REXML::Document.new(open(url, {:proxy => true}).read).root

    error = @xml_data.elements['error']
    raise SteamCondenserError, error.text unless error.nil?

    boards_data.elements.each('entries/entry') do |entry_data|
      if entry_data.elements['steamid'].text.to_i == steam_id
        return GameLeaderboardEntry.new entry_data, self
      end
    end

    nil
  end

  # Returns an array of entries on this leaderboard for the user with the given
  # SteamID and his/her friends
  #
  # @param [Fixnum, SteamId] steam_id The 64bit SteamID or the `SteamId` object
  #        of the user
  # @return [Array<GameLeaderboardEntry>] The entries of the user and his/her
  #         friends
  # raise [SteamCondenserException] if an error occurs while fetching the
  #         leaderboard
  def entry_for_steam_id_friends(steam_id)
    steam_id = steam_id.steam_id64 if steam_id.is_a? SteamId

    url = "#{@url}&steamid=#{steam_id}"
    xml = REXML::Document.new(open(url, {:proxy => true}).read).root

    error = @xml_data.elements['error']
    raise SteamCondenserError, error.text unless error.nil?

    entries = []
    boards_data.elements.each('entries/entry') do |entry_data|
      rank = entry_data.elements['rank'].text.to_i
      entries[rank] = GameLeaderboardEntry.new entry_data, self
    end

    entries
  end

  # Returns the entries on this leaderboard for a given rank range
  #
  # The range is inclusive and a maximum of 5001 entries can be returned in a
  # single request.
  #
  # @param [Fixnum] first The first entry to return from the leaderboard
  # @param [Fixnum] last The last entry to return from the leaderboard
  # @return [Array<GameLeaderboardEntry>] The entries that match the given rank
  #         range
  # raise [SteamCondenserException] if an error occurs while fetching the
  #         leaderboard
  def entry_range(first, last)
    if last < first
      raise SteamCondenserException,
        'First entry must be prior to last entry for leaderboard entry lookup.'
    end

    if (last - first) > 5000
      raise SteamCondenserException,
        'Leaderboard entry lookup is currently limited to a maximum of 5001 entries per request.'
    end

    url = "#{@url}&start=#{first}&end=#{last}"
    xml = REXML::Document.new(open(url, {:proxy => true}).read).root

    error = @xml_data.elements['error']
    raise SteamCondenserError, error.text unless error.nil?

    entries = []
    boards_data.elements.each('entries/entry') do |entry_data|
      rank = entry_data.elements['rank'].text.to_i
      entries[rank] = GameLeaderboardEntry.new entry_data, self
    end

    entries
  end

  private

  # Creates a new leaderboard instance with the given XML data
  #
  # @param [REXML::Element] board_data The XML data of the leaderboard
  def initialize(board_data)
    @url          = board_data.elements['url'].text
    @id           = board_data.elements['lbid'].text.to_i
    @name         = board_data.elements['name'].text
    @entry_count  = board_data.elements['entries'].text.to_i
    @sort_method  = board_data.elements['sortmethod'].text.to_i
    @display_type = board_data.elements['displaytype'].text.to_i
  end

  # Loads the leaderboards of the specified games into the cache
  #
  # @param [String] game_name The short name of the game
  # @raise [SteamCondenserException] if an error occurs while fetching the
  #         leaderboards
  def self.load_leaderboards(game_ame)
    url = "http://steamcommunity.com/stats/#{game_name}/leaderboards/?xml=1"
    boards_data = REXML::Document.new(open(url, {:proxy => true}).read).root

    error = @xml_data.elements['error']
    raise SteamCondenserError, error.text unless error.nil?

    @@leaderboards[game_name] = []
    boards_data.elements.each('leaderboard') do |board_data|
      leaderboard = GameLeaderboard.new board_data
      @@leaderboards[game_name][leaderboard.id] = leaderboard
    end
  end

end
